hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java

```diff
@@ -92,60 +81,47 @@ static boolean isAtMost(JavaMajorVersion currentVersion, JavaMajorVersion maxVer
         return currentVersion.getMajorVersion() <= maxVersion.getMajorVersion();
     }
 
-    private static JavaVersion valueOf(int majorVersion) {
-        for (JavaVersion version : values()) {
-            if (version.majorVersion == majorVersion) {
-                return version;
-            }
-        }
-
-        return null;
-    }
-
+    /**
+     * <pre>
+     * <= JDK 8
+     *     VNUM  := 1\.[1-9]\.[0-9](_[0-9]{2})?         Version number
+     *     PRE   := [a-zA-Z0-9]+                        Pre-release identifier
+     *     java.version
+     *           := $VNUM(-$PRE)*
+     * >= JDK 9
+     *     VNUM  := [1-9][0-9]*(\.(0|[1-9][0-9]*))*     Version number
+     *     PRE   := [a-zA-Z0-9]+                        Pre-release identifier
+     *     BUILD := 0|[1-9][0-9]*                       Build number
+     *     OPT   := [-a-zA-Z0-9\.]+                     Build information
+     *     java.version
+     *           := $VNUM(-$PRE)?\+$BUILD(-$OPT)?
+     *            | $VNUM-$PRE(-$OPT)?
+     *            | $VNUM(+-$OPT)?
+     * </pre>
+     * @see <a href="https://www.oracle.com/java/technologies/javase/versioning-naming.html">
+     *      J2SE SDK/JRE Version String Naming Convention (<= JDK 8)
+     * @see <a href="https://openjdk.org/jeps/223">
+     *      JEP 223: New Version-String Scheme (>= JDK 9)
+     */
     private static JavaMajorVersion detectCurrentVersion() {
-        final ILogger logger = Logger.getLogger(JavaVersion.class);
-        final Class runtimeClass = Runtime.class;
-        final Class versionClass;
-
-        try {
-            versionClass = Class.forName("java.lang.Runtime$Version");
-        } catch (ClassNotFoundException e) {
-            // if Runtime is present but Runtime.Version doesn't, it's Java8 for sure
-            if (logger.isFineEnabled()) {
-                logger.fine("Detected runtime version: Java 8");
-            }
-            return JAVA_8;
-        }
+        String[] version = System.getProperty("java.version").split("[-+.]");
+        int major = Integer.parseInt(version[version[0].equals("1") ? 1 : 0]);
```

Comment:
Could you add here a problem handling too?
E.g. when for some reason the property is removed or the value is set to an unsupported format

Reply:
And it would be nice to have it covered in tests too.